src/NHibernate/Impl/ExpressionQueryImpl.cs

```diff
@@ -125,6 +125,32 @@ public override IList List()
 			}
 		}
 
+		public override IList<T> List<T>()
```

Comment:
Lack of this override caused to call `Session.List` and this resulted in a lost `collection` parameter. So the query applied to the whole table.

Reply:
For correctness of the class, shouldn't a `void List(IList results)` override be added too? (And same for `CollectionFilterImpl`.)

Reply:
Done. I've also added missing Enumerable ovverrides to `ExpressionQueryImpl` to throw NotSupportedException.